File: Week2/node/flows/searchFlow.js
const { linearSearch, binarySearch } = require("../algorithms/search");
const { ask } = require("../utils/input");
const { parseNumber } = require("../utils/parse");
const { printSearchMenu } = require("../menu/menuPrinter");

async function getTargetFromUser() {
  while (true) {
    const raw = await ask("Enter target number (e.g. 7): ");
    const parsed = parseNumber(raw);
    if (parsed.ok) return parsed.value;
    console.log(`Error: ${parsed.error} Try again.`);
  }
}

async function runSearchFlow(arr) {
  printSearchMenu();
  const choice = await ask("Select search type (1-2) or 0 to go back: ");

  if (choice === "0") return { next: "back" };

  const target = await getTargetFromUser();

  if (choice === "1") {
    console.log("\nArray:", arr);
    console.log(`Linear Search → Target ${target} index:`, linearSearch(arr, target));
    return { next: "done" };
  }

  if (choice === "2") {
    const sorted = [...arr].sort((a, b) => a - b);
    console.log("\nSorted Array (for Binary Search):", sorted);
    console.log(`Binary Search → Target ${target} index:`, binarySearch(sorted, target));
    return { next: "done" };
  }

  console.log("Invalid search option.");
  return { next: "done" };
}

module.exports = { runSearchFlow };

File: Week2/python/flows/search_flow.py
from algorithms.search import linear_search, binary_search
from menu.menu_printer import print_search_menu
from utils.parse import parse_number


def get_target_from_user():
    while True:
        raw = input("Enter target number (e.g. 7): ")
        parsed = parse_number(raw)
        if parsed["ok"]:
            return parsed["value"]
        print(f"Error: {parsed['error']} Try again.")


def run_search_flow(arr):
    print_search_menu()
    choice = input("Select search type (1-2) or 0 to go back: ").strip()

    if choice == "0":
        return

    target = get_target_from_user()

    if choice == "1":
        idx = linear_search(arr, target)
        print("\nArray:", arr)
        print("Result:", "Not found" if idx == -1 else f"Found at index {idx}")
        return

    if choice == "2":
        sorted_arr = sorted(arr)
        idx = binary_search(sorted_arr, target)
        print("\nSorted Array (for Binary Search):", sorted_arr)
        print("Result:", "Not found" if idx == -1 else f"Found at index {idx}")
        return

    print("Invalid search option.")
